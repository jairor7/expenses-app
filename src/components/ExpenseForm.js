import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  state = {
    description: "",
    note: "",
    amount: "",
    createdAt: moment(),
    calendarFocused: false,
    error: "",
  };

  componentDidMount() {
    const { expense } = this.props;
    if (expense) {
      this.setState({
        description: expense.description,
        note: expense.note,
        amount: expense.amount,
        createdAt: moment(expense.createdAt),
      });
    }
  }

  onChangeDescription = (e) => {
    this.setState(() => ({
      description: e.target.value,
    }));
  };

  onChangeNote = (e) => {
    this.setState(() => ({
      note: e.target.value,
    }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,4}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount: amount,
      }));
    }
  };

  onChangeDate = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt,
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused,
    }));
  };

  onSubmit = (e) => {
    const { description, note, amount, createdAt } = this.state;
    e.preventDefault();
    if (!description || !amount) {
      this.setState(() => {
        return {
          error: "Please provide description and amount",
        };
      });
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: description,
        amount: parseFloat(amount, 10),
        createdAt: createdAt.valueOf(),
        note: note,
      });
    }
  };

  render() {
    const { description, note, amount, error } = this.state;
    const { expense } = this.props;
    return (
      <div>
        {error && <p>{error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={description}
            onChange={this.onChangeDescription}
          />
          <input
            type="text"
            value={amount}
            onChange={this.onChangeAmount}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onChangeDate} // PropTypes.func.isRequired
            focused={this.state.calendarFocused} // PropTypes.bool
            onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
            id="calendar-id-1" // PropTypes.string.isRequired,
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={note}
            onChange={this.onChangeNote}
            placeholder="Add a note for yopur expense (opcional)"
          />
          <button>{expense ? "Edit expense" : "Add Expense"}</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
