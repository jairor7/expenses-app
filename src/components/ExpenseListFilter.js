import React from "react";
import { DateRangePicker } from "react-dates";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setEndDate,
  setStartDate,
} from "../store/redux/actions/filters";
import "react-dates/initialize";

export class ExpenseListFilter extends React.Component {
  state = {
    focusedInput: null,
  };
  onChangeValueFilter = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onChangeSelectFilter = (e) => {
    const filterOption = e.target.value;
    if (filterOption === "date") {
      this.props.sortByDate();
    } else if (filterOption === "amount") {
      this.props.sortByAmount();
    }
  };
  onChangeDates = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  render() {
    const { filters } = this.props;
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onChangeValueFilter}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={this.onChangeSelectFilter}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={filters.startDate} // momentPropTypes.momentObj or null,
          startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
          endDate={filters.endDate} // momentPropTypes.momentObj or null,
          endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
          onDatesChange={this.onChangeDates} // PropTypes.func.isRequired,
          focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={(focusedInput) => this.setState({ focusedInput })} // PropTypes.func.isRequired,
          small={true}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
