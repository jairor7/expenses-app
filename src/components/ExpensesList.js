import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../store/redux/selectors/expenses";

export const ExpensesList = (props) => (
  <div>
    <h1>Expenses List</h1>
    <p>This is the expenses list</p>
    {props.expenses.length === 0 ? (
      <p>No items - no expenses</p>
    ) : (
      props.expenses.map((expense) => (
        <ExpenseListItem key={expense.id} {...expense} />
      ))
    )}
  </div>
);

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state),
});

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(ExpensesList);
