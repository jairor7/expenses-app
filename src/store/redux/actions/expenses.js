import { v4 as uuid } from "uuid";
import { expenseTypes } from "../types";

// ADD_EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0,
} = {}) => ({
  type: expenseTypes.ADD_EXPENSE,
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt,
  },
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: expenseTypes.REMOVE_EXPENSE,
  id,
});

// EDIT_EXPENSE
export const editExpense = ({ id, ...updates } = {}) => ({
  type: expenseTypes.EDIT_EXPENSE,
  id,
  updates,
});
