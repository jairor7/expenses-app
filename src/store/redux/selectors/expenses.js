// import moment from "moment";
// Get visible expenses
// eslint-disable-next-line import/no-anonymous-default-export
export default ({ expenses, filters }) => {
  const { text, sortBy, startDate, endDate } = filters;
  return expenses
    .filter((expense) => {
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(expense.createdAt, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(expense.createdAt, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt <= b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount <= b.amount ? 1 : -1;
      } else {
        return false;
      }
    });
};
