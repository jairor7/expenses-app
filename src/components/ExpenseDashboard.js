import React from "react";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpensesList from "./ExpensesList";

const ExpenseDashboard = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
  </div>
);

export default ExpenseDashboard;
