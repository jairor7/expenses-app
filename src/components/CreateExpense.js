import React from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../store/redux/actions/expenses";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";

export const CreateExpense = (props) => {
  const navigate = useNavigate();
  const onSubmit = (expense) => {
    props.addExpense({
      description: expense.description,
      amount: parseFloat(expense.amount, 10),
      createdAt: expense.createdAt.valueOf(),
      note: expense.note,
    });
    navigate("/");
  };
  return (
    <div>
      <h1>Create Expense</h1>
      <ExpenseForm onSubmit={onSubmit} />
    </div>
  );
};

const mapStateToProps = (state) => {
  return {};
};

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateExpense);
