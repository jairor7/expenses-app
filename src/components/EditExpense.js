import React from "react";
import { connect } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import { editExpense } from "../store/redux/actions/expenses";
import { removeExpense } from "../store/redux/actions/expenses";
import ExpenseForm from "./ExpenseForm";

export const EditExpense = (props) => {
  const { id } = useParams();
  const navigate = useNavigate();
  const expense = props.expenses.find((expense) => expense.id === id);

  const onSubmit = (expense) => {
    props.editExpense({
      id,
      description: expense.description,
      amount: parseFloat(expense.amount, 10),
      createdAt: expense.createdAt.valueOf(),
      note: expense.note,
    });
    navigate("/");
  };

  const onClickRemoveExpense = () => {
    props.removeExpense({ id });
    navigate("/");
  };

  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm onSubmit={onSubmit} expense={expense} />
      <button onClick={onClickRemoveExpense}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (expense) => dispatch(editExpense(expense)),
    removeExpense: (id) => {
      dispatch(removeExpense(id));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
