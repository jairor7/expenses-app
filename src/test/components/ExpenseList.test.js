import React from "react";
import { shallow } from "enzyme";
import { ExpensesList } from "../../components/ExpensesList";
import { expenses } from "../fixtures/expenses";
import toJSON from "enzyme-to-json";

test("should render Expenselist with expenses", () => {
  const wrapper = shallow(<ExpensesList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render ExpenseList with empty message", () => {
  const wrapper = shallow(<ExpensesList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
