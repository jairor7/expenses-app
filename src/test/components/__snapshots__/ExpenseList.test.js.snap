// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <h1>
    Expenses List
  </h1>
  <p>
    This is the expenses list
  </p>
  <p>
    No items - no expenses
  </p>
</div>
`;

exports[`should render Expenselist with expenses 1`] = `
<div>
  <h1>
    Expenses List
  </h1>
  <p>
    This is the expenses list
  </p>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
  <ExpenseListItem
    amount={195}
    createdAt={777600000}
    description="Gum"
    id="4"
    key="4"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={604800000}
    description="Credit Card"
    id="5"
    key="5"
    note=""
  />
</div>
`;
